### PR TITLE
Issue 15243 - rejects-valid on variadic

### PR DESCRIPTION
**Pull request by 9rnsr**

### CODE REVIEW

File: src/dtemplate.d [hunk 1]
```diff
@@ -3433,28 +3433,48 @@ extern (C++) MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplatePara
...
+                    // Bugzilla 15243
+                    if (!reliesOnTident(fparam.type, parameters, inferStart))
+                    {
+                        Type tx = fparam.type.semantic(Loc(), sc);
+                        if (tx.ty == Terror)
+                        {
+                            result = MATCHnomatch;
+                            return;
+                        }
+                        fparam.type = tx;
```
Comment (WalterBright) [inline src/dtemplate.d]: Shouldn't `nfparams` get updated here?

File: src/dtemplate.d [hunk 2]
```diff
@@ -3433,28 +3433,48 @@ extern (C++) MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplatePara
...
+                    }
                 }
-                //printf("\t-> this   = %d, ", t->ty); t->print();
-                //printf("\t-> tparam = %d, ", tparam->ty); tparam->print();
+                //printf("\t. this   = %d, ", t.ty); t.print();
+                //printf("\t. tparam = %d, ", tparam.ty); tparam.print();
+
+                // Update for newly found TypeTuple parameter.
+                nfparams = Parameter.dim(tp.parameters);
```
Comment (WalterBright) [inline src/dtemplate.d]: and not here?

File: src/dtemplate.d (hunk 1 above)
Comment (9rnsr) [inline src/dtemplate.d]: Thanks. From your point out, I found and got more thoughts:
In here, the original `fparam.type` never contains `Ttuple`, so we don't have to use `Parameter.dim(tp.parameters)` to iterate `tp.parameters`. Therefore we can calculate `nfparams` only once after this `for` loop.

I updated the code based on my thought.

**Comment by JinShil** [comment]
Revived at #7403

**Pull request closed without merging**